com.ibm.streamsx.topology/opt/python/include/splpy_cr.h

```diff
@@ -0,0 +1,281 @@
+  template<>
+  struct Allocator<PyObject *>
+  {
+    // allocate: seems to be no good way.   It seems not to be needed anyway.
+
+    static void deallocate(PyObject * t) {
+      // To deallocate PyObject * , DECREF it rather than deleting it.
+      streamsx::topology::SplpyGIL gil;
+      Py_XDECREF(t);
+    }
+  };
+
+  template<typename T> struct Referencer;
+  template<>
+  struct Referencer<PyObject *>
+  {
+    // We don't allow PyObject * to be dereferenced, so the dereference
+    // type is PyObject *.
+    typedef PyObject * dereference_t;
+    typedef PyObject * reference_t;
+    static PyObject * dereference(PyObject * t) { return t; }
+    static PyObject const * dereference(PyObject const * t) { return t; }
+    static PyObject * reference(PyObject * t) { return t; }
+    static PyObject const * reference(PyObject const * t) { return t; }
+  };
+
+
+  ByteBuffer<Checkpoint> & operator<<(ByteBuffer<Checkpoint> & ckpt, PyObject * obj){
+    using namespace streamsx::topology;
+
+    SPLAPPTRC(L_TRACE, "operator << (ByteBuffer<Checkpoint> &, PyObject *): enter", "python");
+
+    SPL::blob bytes;
+    {
+      SplpyGIL gil;
+
+      // TODO ideally we don't load dumps every time.
+      PyObject * dumps = SplpyGeneral::loadFunction("dill", "dumps");
+      PyObject * args = PyTuple_New(1);
+      Py_INCREF(obj);
+      PyTuple_SET_ITEM(args, 0, obj);
+      PyObject * ret = PyObject_CallObject(dumps, args);
+      if (ret == NULL) {
+        SplpyGeneral::tracePythonError();
+        Py_DECREF(dumps);
+        Py_DECREF(args);
+        throw SplpyGeneral::pythonException("dill.dumps");
+      }
+      Py_DECREF(dumps);
+      Py_DECREF(args);
+      // ret is the dilled object
+      pySplValueFromPyObject(bytes, ret);
+      Py_DECREF(ret);
+    }
+    ckpt << bytes;
+
+    SPLAPPTRC(L_TRACE, "operator << (ByteBuffer<Checkpoint>&, PyObject *): exit", "python");
+    return ckpt;
+  }
+
+  ByteBuffer<Checkpoint> & operator>>(ByteBuffer<Checkpoint> & ckpt, PyObject * &obj){
+    using namespace streamsx::topology;
+    SPLAPPTRC(L_TRACE, "operator >> (ByteBuffer<Checkpoint>&, PyObject *&): enter", "python");
+    SPL::blob bytes;
+    ckpt >> bytes;
+
+    SplpyGIL gil;
+
+    // TODO ideally we don't load loads every time.
+    PyObject * loads = SplpyGeneral::loadFunction("dill", "loads");
+    PyObject * args = PyTuple_New(1);
+    PyObject * pickle = pySplValueToPyObject(bytes);
+    PyTuple_SET_ITEM(args, 0, pickle);
+    PyObject * ret = PyObject_CallObject(loads, args);
+    if (!ret) {
+      SplpyGeneral::tracePythonError();
+      Py_DECREF(loads);
+      Py_DECREF(args);
+      throw SplpyGeneral::pythonException("dill.loads");
+    }
+    Py_DECREF(loads);
+    Py_DECREF(args);
+    obj = ret;
+
+    SPLAPPTRC(L_TRACE, "operator >> (ByteBuffer<Checkpoint>&, PyObject *&): exit", "python");
+
+    return ckpt;
+  }
+}
+
+std::ostream & operator <<(std::ostream &ostr, PyObject * obj){
```

Comment:
What's this method for? Some explanation would be useful as why it's needed and why it doesn't handle conversion to strings from objects using existing methods?

Reply:
Comments added.  The window library uses it for diagnostic messages.